Editing

m-elseberg opened the pull request:
Editing capabilities are, of course, a central part of many GIS apps. However, the current implementation is limited to polygons and the OGC Features API.

This pull request suggests a more general-purpose editing solution, which evolved throughout one of our projects. Features include:

- creating and updating any geometry type (polygons, line strings, points, circles),
- pre-configurable or fully custom attribute forms,
- ability to save features anywhere,
- snapping.

To test this implementation, open the following local sample: <http://localhost:5173/samples/editing-sample/>. For ease of testing, this sample saves the created features to the browser's IndexedDB.

This solution is not expected to be feature-complete or to meet everyone's needs. For one thing, it is still lacking documentation.

(screenshots: [1](https://github.com/user-attachments/assets/7ff9fcd0-eede-4902-9a69-38b2ba40d4b6), [2](https://github.com/user-attachments/assets/55b736fe-1ac5-4f48-993f-d447d021ae46))

m-elseberg committed:
Add new-editing package

m-elseberg committed:
Add editing-sample

m-elseberg committed:
Patch react-color@2.19.3

Gets rid of two console error messages:
- "Warning: Circle2: Support for defaultProps will be removed from function components in a future major release. Use JavaScript default parameters instead."
- "Warning: CircleSwatch2: Support for defaultProps will be removed from function components in a future major release. Use JavaScript default parameters instead."

See: https://github.com/casesandberg/react-color/issues/888

m-elseberg committed:
Run prettier

mbeckem committed:
Build sample app

mbeckem committed:
Update license config

mbeckem commented:
I updated the vite config and license report config to get the samples to build. Nitpick: the react-color dependency looks unmaintained and references some very old packages (e.g. `@material/icons`). Should probably use an alternative.

mbeckem committed:
Minor changes during code review

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem commented:
Okay, i've finally had a chance to look at your PR. First of all, the end user experience seems to be (almost) ideal, great job.

Looking at the implementation, i've got some feedback to share with you.
Not everything is extremely important, but some of it is: I feel that it is important to start out with a high quality implementation (also maintenance-vise) because this package will probably be quite heavily used.

- Tests are missing missing.

- Really needs comments, at least on all exported entities and their properties.

- Are all exports necessary? I think the public interface could be slimmed down a bit without losing functionality.

- Ensure nice behavior with dynamic size (no fixed size or useless scrollbars)

- Icon Buttons should probably have a tooltip (could be same as aria label).

- Eliminate FieldInputsProvider and create a `FormContent` concept (FieldInput[] or completely custom Form). This can also replace the `children` API.

- ColorInput is difficult to support well and should be removed for the time being.
    - Note that chakra v3 has a color input widget so it might be possible to re-add this feature at a later time.

- Idea: CustomInput type that supports free form rendering, i.e. it renders some react component within the form.
  This way, arbitrary input types can be added by developers (including ColorInput).
  Requires a render function (might have to pass some props in for aria stuff, form ids etc.).

- Union variants should be exported if the unions are exported as well. For example `InitialStep` as a developer might want to name the type in their code.

- For consistency with other packages:
    - If raw OpenLayers layers appear in APIs, use `olLayer` instead of `layer` to avoid confusion
    - Or (even better, if possible), use the `Layer` type (or other fitting types) from the map package instead

- I like the different *Handler functions.
  Their signatures would have to be changed in the future if we ever get around creating our own `Feature` class.
  Currently, we have `BaseFeature` in the map package (a barebones interface that combines attributes with geometry and projection)
  but we would like to create a fully featured `Feature` class at some point (as a reactive Model, possibly using OL's `Feature` internally).

- PropertyFormContext _might_ be a performance problem in the future.
  In react, changing the context value requires rerendering all children that use that context.
  Combined with react-style immutability this can be a performance hazard.
  The reactivity API could be used to make a `FormModel` available to all children, that would not this problem (because of fine grained mutability, the model would stay the same but small details would change).

- Not sure why `FeatureTemplate` can be considered an `Action`, this feels wrong.
  Better: `{kind: "create", template: FeatureTemplate}` ?
  This also allows for easy future extensions with more action types.

- Not a fan of the `tsx` directory, I think that directory level can be removed.

- Please don't use console.* for logging, use the `Logger` type from `@open-pioneer/core` instead.
  You can simplify create a module level constant, like `const LOG = createLogger("map:AbstractLayer");`.

- I _think_ using the reactivity API for the editing actions and the undo management might be easier than the current approach.

- Nitpick: I think these type aliases hurt readability more than they help.

  ```ts
  export type Callback<T> = (newValue: T) => void;
  export type VoidCallback = () => void;
  export type AsyncVoidCallback = () => Promise<void>;
  ```

- Nitpick: camel case for input types (e.g. textArea) feels wrong (-> "text-area" instead)

- Its nice that this works but it's pretty risky to rely on it.
  The `this` parameter has not been bound explicitly (by us), although it _might_ be a feature of FormatJS:
  ```
  const { formatMessage } = useIntl();
  ```

- Just a tip: take a look at [`useEvent`](https://github.com/open-pioneer/trails-core-packages/blob/a208b32e3c862c45319b85939f18b562062a29d3/src/packages/react-utils/useEvent.ts), it can help simplifying some usages of `useCallback` (no need to track dependencies).

m-elseberg committed:
Merge remote-tracking branch 'upstream/main' into new-editing

Conflicts:
	package.json
	pnpm-lock.yaml
	pnpm-workspace.yaml
	support/license-config.yaml
	vite.config.ts

All conflicts resolved using upstream/main.

m-elseberg committed:
Skip pre-commit hook and workflows (to be reverted)

m-elseberg committed:
Fix project state

- Remove old dependencies
- Replace @open-pioneer/chakra-integration imports by @chakra-ui/react
- Auto-fix headers/header-format linter errors
- Silence @typescript-eslint/no-non-null-assertion errors for now

m-elseberg committed:
Let Monsieur Claude upgrade us to Chakra 3

Claude Sonnet v4.5

m-elseberg committed:
Fix remaining React/Chakra/OPT issues

- Wait for map model to be loaded
- Fix Table.Root variant
- Supply explicit initial value to useRef()
- Run prettier on AI code changes

m-elseberg committed:
Rewrite sample app (no feature persistence)

m-elseberg committed:
Don't update layer from inside editing package

m-elseberg committed:
Rewrite color picker

m-elseberg committed:
Eliminate intermediate 'tsx' directory

m-elseberg committed:
Apply minor code changes

- Use kebab case for field input types
- Redefine Action type
- Use olLayer parameter name for EditingHandler
- Remove console logging

m-elseberg committed:
Eliminate callback types

m-elseberg committed:
Fix minor Chakra UI issues

- Add required and select indicators
- Keep left and right input control borders visible
- Fix red buttons and avoid horizontal scroll bar
- Fix confirmation dialog layout

m-elseberg committed:
Revert "Skip pre-commit hook and workflows (to be reverted)"

This reverts commit cc033138398670e7316cf0ead5a8b9b7907d6b13.

m-elseberg committed:
Apply minor fixes

m-elseberg committed:
Rewrite core editing logic using Reactivity API

m-elseberg committed:
Adjust headings, logging, and type exports

- Replace Heading by TitledSection
- Add logging to editing callbacks
- Export union member types

m-elseberg commented:
Current progress:

> * [ ] Tests are missing.
> * [ ] Really needs comments, at least on all exported entities and their properties.
> * [ ] Are all exports necessary? I think the public interface could be slimmed down a bit without losing functionality.
> * [x] Ensure nice behavior with dynamic size (no fixed size or useless scrollbars)
> * [ ] Icon Buttons should probably have a tooltip (could be same as aria label).
> * [ ] Eliminate FieldInputsProvider and create a `FormContent` concept […].
> * [x] ColorInput is difficult to support well and should be removed for the time being. […]
> * [ ] Idea: CustomInput type that supports free form rendering […].
> * [x] Union variants should be exported if the unions are exported as well. […]
> * [ ] For consistency with other packages: […]
> * [ ] I like the different *Handler functions. […]
> * [ ] PropertyFormContext _might_ be a performance problem in the future. […]
> * [x] Not sure why `FeatureTemplate` can be considered an `Action`, this feels wrong. […]
> * [x] Not a fan of the `tsx` directory, I think that directory level can be removed.
> * [x] Please don't use console.* for logging, use the `Logger` type from `@open-pioneer/core` instead. […]
> * [x] I _think_ using the reactivity API for the editing actions and the undo management might be easier than the current approach.
> * [x] Nitpick: I think these type aliases hurt readability more than they help. […]
> * [x] Nitpick: camel case for input types (e.g. textArea) feels wrong (-> "text-area" instead)
> * [ ] Just a tip: take a look at [`useEvent`](https://github.com/open-pioneer/trails-core-packages/blob/a208b32e3c862c45319b85939f18b562062a29d3/src/packages/react-utils/useEvent.ts), it can help simplifying some usages of `useCallback` […].
> * [ ] Icons should use lucide (`react-icons/lu`) wherever possible
> * [x] Sample App can be simplified (too much code). Persistence isn't required.
> * [x] Use TitledSection/SectionHeading instead of raw chakra `Heading`. […]
> * [ ] PropertyFormContext can be made into a model using the reactivity API for better performance and less `useCallback` noise.
> * [x] useEditing should be refactored into a plain JavaScript class (or multiple ones). […]

m-elseberg committed:
Rerun GitHub actions (empty commit)

m-elseberg committed:
Rename some identifiers

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

m-elseberg committed:
Use reactivity API for property form context

m-elseberg committed:
Add tooltips

m-elseberg committed:
Prefer Lucide icons

m-elseberg committed:
Use 'olLayer' identifiers in public API

m-elseberg committed:
Prefer useEvent()

m-elseberg committed:
Fix property updates, notifier position, and delete button hover effect

m-elseberg committed:
Eliminate types.ts

m-elseberg committed:
Refine new-editing package

Documentation and testing:
- Add readme and license file
- Add doc comments
- Add unit tests
- Add license configuration for Phosphor icons

Features:
- Support custom input controls with render() function
- Support dynamic form templates with renderForm() function
- Add combo box, switch, and radio button group controls
- Add property functions
- Add Header component
- Make notifier display duration configurable

Refactoring:
- Restructure package
- Replace OpenLayers layer types by OpenPioneer layers
- Remove 'children' API from Editor
- Replace FieldInputsProvider interface by FormTemplateProvider function
- Rewrite input controls
- Rename various identifiers

Bug fixing:
- Fix null date handling
- Increase save button width for German locale

m-elseberg committed:
Merge remote-tracking branch 'upstream/main' into new-editing

m-elseberg committed:
Fix unit test and adjust 'key' prop

Fixes:
- "An update to RadioGroupRoot inside a test was not wrapped in act(...)."
- "Each child in a list should have a unique 'key' prop."

m-elseberg committed:
Slightly adjust documentation and sample app

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem committed:
Update lockfile

mbeckem committed:
Configure package for building; start review

mbeckem committed:
Rename package

mbeckem committed:
Fix timing problem

antoniave committed:
Use chakra snippet for RadioGroup

antoniave committed:
simplify class hierarchy

mbeckem committed:
Move features from hooks to implementation

mbeckem committed:
Move editing callbacks into context; rename directories

mbeckem committed:
Rename step ids (consistent with type names) and make callback signatures future proof

antoniave committed:
use createElement instead of calling react component as function

mbeckem committed:
Change pnpm-lock to fix cache (?)

mbeckem committed:
???

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem committed:
Remove overrides

mbeckem committed:
Remove redundant package lock file (?!)

mbeckem committed:
Update lockfile

antoniave committed:
restrict and document allowed geometry types

mbeckem committed:
Review public interface

- Make useEditing / useGeometryEditing internal for now
- Rename `editingHandler: EditingHandler` to `storage: EditingStorage`
- Rename `formTemplateResolver` to `resolveFormTemplate`
- Inline typings for `resolveFormTemplate` and `onEditingStepChange`

mbeckem committed:
Fix docs

mbeckem committed:
Use 'false' to hide notifications instead of timeout 0

mbeckem committed:
Fix typedoc warnings

antoniave committed:
move interfaces and types to the top

mbeckem committed:
Use groups to organize typedoc entities

mbeckem committed:
attributes --> properties

mbeckem committed:
Rename prototype --> defaultAttributes to avoid collisions with builtin javascript features

mbeckem committed:
Use new highlights API

mbeckem committed:
Integrate feedback

- Remove non-essential properties from example in README
- `get propertiesObject` --> `getPropertiesAsObject()`

mbeckem committed:
Clarify documentation

mbeckem committed:
Hide non-technical error messages from the UI by default. Use StorageError instead to show a custom string to the user.

mbeckem committed:
Remove widget title from editing form

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem committed:
Wrap component with a box and classname

mbeckem committed:
Add css classes to action selector components

mbeckem committed:
Add a hint about the react node's context

antoniave committed:
use topmost layer instead of raw OL layer

antoniave committed:
fix tests

mbeckem committed:
Assign css classes to property form components

mbeckem committed:
Merge remote-tracking branch 'origin/new-editing' into new-editing

mbeckem committed:
Rename package to 'feature-editing'

mbeckem committed:
Remove useGeometryEditing docs from README (hook is private for now)

mbeckem committed:
Rename Editor.tsx -> FeatureEditor.tsx

mbeckem committed:
Use 'Layer' instead of 'SimpleLayer' for public API

mbeckem committed:
Add npm badge

mbeckem committed:
Rename EditingStorage -> FeatureWriter

antoniave committed:
cleanup package rename; rename sample folder

antoniave committed:
added tooltip hints for interactions

antoniave committed:
fix width and height on small screens (sample app)

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem committed:
Describe data flow

antoniave committed:
improve drawLayer cleanup

mbeckem committed:
Update test app

- inline FeatureWriter implementation
- add some comments

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

mbeckem committed:
Fix lockfile

antoniave committed:
document which layer types can be used

mbeckem committed:
Add a self contained example for the data flow section

mbeckem committed:
Remove copyright header from example

mbeckem committed:
fix typo

arnevogt committed:
fix overlay example in read me

antoniave committed:
document supported keyboard shortcuts

mbeckem committed:
Update to latest dev snapshots

mbeckem committed:
Rename module

arnevogt committed:
Support dark mode in components (#596)

antoniave committed:
improve accessibility

mbeckem committed:
Fix lockfile

mbeckem committed:
Review tests

mbeckem committed:
Merge remote-tracking branch 'upstream/main' into new-editing

arnevogt committed:
adjustments for dark mode

arnevogt committed:
Merge remote-tracking branch 'm-elseberg/new-editing' into new-editing-m-elseberg

mbeckem committed:
a11y fine tuning

mbeckem committed:
Add changeset
